### PR TITLE
Useorm

### PR DESCRIPTION
**Pull request by TordJoranger**
Changes include:

Refactored to use nullable enable.

Removed PcsTopic enum, changed to use the string values of the events for now.

Changed the way we query DB for messages based on id. We still use custum sql, but we map it into strongly typed types.
I've chosen Dapper as an ORM for this due to its simplicity in writing custum SQL. Equivalent to hbm.xml in Main.

Created version 1 interface for all our events. consumers of the servicebus should implement interface to deserialize into.

Created unit tests for serializing and deserializing said events.

created unit tests that seals the interfaces. 

Minor refactorings here and there, including some changes to gitignore and dictionary.

### CODE REVIEW

File: src/Equinor.ProCoSys.BusReceiver/BusReceiverService.cs [hunk 1]
```diff
@@ -1,11 +1,12 @@
-﻿using Equinor.ProCoSys.PcsServiceBus;
+﻿
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusReceiver/BusReceiverService.cs:1]: Blank line

File: src/Equinor.ProCoSys.BusReceiver/BusReceiverService.cs [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+using Equinor.ProCoSys.PcsServiceBus;
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusReceiver/BusReceiverService.cs:2]: Not in use

File: src/Equinor.ProCoSys.BusSender.Core/Mappers/BusEventToTopicMapper.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Equinor.ProCoSys.BusSenderWorker.Core.Mappers;
+
+/***
+ * If and when we merge topics, we will use this class to map.
+    ie,  if(e == 'tag') return PcsTopicConstants.Completion.
+ */
+internal static class BusEventToTopicMapper
+{
+    public static string Map(string e) => e;
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusSender.Core/Mappers/BusEventToTopicMapper.cs:9]: What is 'e'?

File: src/Equinor.ProCoSys.BusSender.Core/Models/McPkgMilestoneEvent.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using Equinor.ProCoSys.PcsServiceBus;
+using Equinor.ProCoSys.PcsServiceBus.Interfaces;
+#pragma warning disable CS8618
+
+namespace Equinor.ProCoSys.BusSenderWorker.Core.Models;
+
+public class McPkgMilestoneEvent : IMcPkgMilestoneEventV1
+{
+    public string Plant { get; init; }
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusSender.Core/Models/McPkgMilestoneEvent.cs:10]: init is interesting, can we use it on all xxxEvent classes?

File: src/Equinor.ProCoSys.BusSender.Core/Services/BusSenderService.cs [hunk 5]
```diff
@@ -80,27 +84,27 @@ private async Task ProcessBusEvents(List<BusEvent> events)
         var unProcessedEvents = events.Where(busEvent => busEvent.Status == Status.UnProcessed).ToList();
         _logger.LogInformation("Amount of messages to process: {count} ", unProcessedEvents.Count);
 
-        foreach (var e in unProcessedEvents.Where( e => IsSimpleMessage(e) || e.Event == TagTopic.TopicName))
+        foreach (var e in unProcessedEvents.Where(e => IsSimpleMessage(e) || e.Event == TagTopic.TopicName))
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusSender.Core/Services/BusSenderService.cs:87]: e -> simpleUnprocessedTagEvent ?

File: src/Equinor.ProCoSys.BusSender.Core/Services/BusSenderService.cs [hunk 6]
```diff
@@ -173,25 +179,28 @@ private static void SetAllButOneEventToSkipped(IEnumerable<BusEvent> group)
...
-        IEnumerable<BusEvent> events)
+    private static IEnumerable<IGrouping<(string Event, string Message), BusEvent>>
+        FilterOnSimpleMessagesAndGroupDuplicates(
+            IEnumerable<BusEvent> events)
         => events.Where(IsSimpleMessage)
             .GroupBy(e => (e.Event, e.Message));
 
-    private static bool IsSimpleMessage(BusEvent e) 
-        => long.TryParse(e.Message, out _) 
+    private static bool IsSimpleMessage(BusEvent e)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusSender.Core/Services/BusSenderService.cs:188]: "BusEvent busEvent" seems to be the pattern in other methods in the class.

File: src/Equinor.ProCoSys.BusSender.Core/Services/BusSenderService.cs [hunk 7]
```diff
@@ -211,7 +220,7 @@ private async Task UpdateEventBasedOnTopic(BusEvent busEvent)
                     busEvent.Message = await _service.AttachTagDetails(busEvent.Message);
                     break;
                 }
-            case ChecklistTopic.TopicName:
+            case "checklist":
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusSender.Core/Services/BusSenderService.cs:223]: Use constant?

File: src/Equinor.ProCoSys.BusSender.Core/Interfaces/IDapperRepository.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+using System.Threading.Tasks;
+using Equinor.ProCoSys.PcsServiceBus.Interfaces;
+
+namespace Equinor.ProCoSys.BusSenderWorker.Core.Interfaces;
+
+public interface IDapperRepository
+{
+    Task<IEnumerable<T>> Query<T>(string queryString, string? objectId) where T : IHasEventType;
+    Task<T?> QuerySingle<T>(string queryString, string objectId) where T : IHasEventType;
+}
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Core/Interfaces/IDapperRepository.cs:11]: Is this a good and consistent name? Repository naming should reflect the objects you are working on, not the infrastructure/technology?

File: src/Equinor.ProCoSys.BusSender.Core/Models/ActionEvent.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    public string? ActionByRole { get; set; }
+    public Guid? ActionByRoleGuid { get; set; }
+    public Guid? ResponsibleOid { get; set; }
+    public string? ResponsibleRole { get; set; }
+    public Guid? ResponsibleRoleGuid { get; set; }
+    public DateTime LastUpdated { get; set; }
+    public DateTime? SignedAt { get; set; }
+    public Guid? SignedBy { get; set; }
+    
+    public string EventType => PcsEventConstants.ActionCreateOrUpdate;
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Core/Models/ActionEvent.cs:39]: Is this an event og dto? In either case we should not postfix with the category name?

File: src/Equinor.ProCoSys.BusSender.Core/Models/BusEventMessage.cs [hunk 10]
```diff
@@ -1,9 +1,10 @@
 ﻿namespace Equinor.ProCoSys.BusSenderWorker.Core.Models;
+#pragma warning disable CS8618
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Core/Models/BusEventMessage.cs:3]: Why is this here? Do we need some explanaition here?

File: src/Equinor.ProCoSys.BusSender.Core/Telemetry/ApplicationInsightsTelemetryClient.cs [hunk 11]
```diff
@@ -37,7 +37,7 @@ public ApplicationInsightsTelemetryClient(TelemetryConfiguration telemetryConfig
             .GetMetric(name, dimension1Name)
             .TrackValue(metric, dimension1Value);
 
-    public void TrackMetric(string name, double metric, string dimension1Name, string dimension2Name, string dimension1Value, string dimension2Value) =>
+    public void TrackMetric(string name, double metric, string dimension1Name, string dimension2Name, string? dimension1Value, string dimension2Value) =>
         _ai
             .GetMetric(name, dimension1Name, dimension2Name)
             .TrackValue(metric, dimension1Value, dimension2Value);
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Core/Telemetry/ApplicationInsightsTelemetryClient.cs:43]: Use tuples instead of numbered properties?

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateOnlyHandler.cs [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public static readonly DateOnlyHandler Default = new();
+
+    public override void SetValue(IDbDataParameter parameter, DateOnly value) 
+        => parameter.Value = value.ToString("yyyy-MM-dd");
+
+    public override DateOnly Parse(object value)
+    {
+        var dateTime = DateTime.SpecifyKind((DateTime)value, DateTimeKind.Utc);
+        return new DateOnly(dateTime.Year, dateTime.Month, dateTime.Day);
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateOnlyHandler.cs:17]: Validation needed before casting value. DateTime.TryParse?

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateTimeUtcHandler.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+internal class DateTimeUtcHandler : SqlMapper.TypeHandler<DateTime>
+{
+    public static readonly DateTimeUtcHandler Default = new();
+    public override void SetValue(IDbDataParameter parameter, DateTime value) 
+        => parameter.Value = value;
+
+    public override DateTime Parse(object value) 
+        => DateTime.SpecifyKind((DateTime)value, DateTimeKind.Utc);
+}
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateTimeUtcHandler.cs:15]: TryParse before casting?

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    private readonly ILogger<DapperRepository> _logger;
+
+    public DapperRepository(BusSenderServiceContext context, ILogger<DapperRepository> logger)
+    {
+        _context = context;
+        _logger = logger;
+    }
+
+    public async Task<IEnumerable<T>> Query<T>(string queryString, string? objectId) where T : IHasEventType
+    {
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs:26]: Is objectId nullable?

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            await _context.Database.OpenConnectionAsync();
+        }
+
+        var events = connection.Query<T>(queryString).ToList();
+        if (events.Count == 0)
+        {
+            _logger.LogError("Object/Entity with id {objectId} did not return anything", objectId);
+            return default;
+        }
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs:55]: The return value can be T not null?

File: src/Equinor.ProCoSys.PcsBus/Queries/CallOffQuery.cs [hunk 16]
```diff
@@ -2,41 +2,72 @@
...
+        //'"", ""IsVoided"" : ' || decode(co.isVoided,'Y', 'true', 'N', 'false') ||
+        //', ""CreatedAt"" : ""' || TO_CHAR(co.createdat, 'yyyy-mm-dd hh24:mi:ss') ||
+        //'""}}' as message
+        //from calloff co
+        //    join purchaseorder po on po.package_id = co.package_id       
+        //    left join responsible r on r.responsible_id = co.responsible_id
+        //    left join library contractor on contractor.library_id = co.contractor_id
+        //    left join library supplier on supplier.library_id = co.supplier_id
+        //{whereClause}";
     }
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.PcsBus/Queries/CallOffQuery.cs:72]: Could remove code commented out.

File: src/Equinor.ProCoSys.PcsBus/Queries/CommPkgOperationQuery.cs [hunk 17]
```diff
@@ -2,32 +2,55 @@
...
+        //    ""PmRoutine"" : ' || decode(co.pmroutine,'Y', 'true', 'N', 'false') || ',
+        //    ""CommissioningResp"" : ' ||  decode(co.commissioningresp,'Y', 'true', 'N', 'false') || ',
+        //    ""ValveBlindingList"" : ""' || decode(co.valveblindinglist,'Y', 'true', 'N', 'false') ||  '"", 
+        //    ""LastUpdated"" : ""' || TO_CHAR(co.LAST_UPDATED, 'yyyy-mm-dd hh24:mi:ss')  || '""
+        //    }}'
+        //from commpkg_operation co
+        //    join commpkg c on c.commpkg_id = co.commpkg_id
+        //    join project p ON p.project_id = c.project_id
+        //{whereClause}";
     }
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.PcsBus/Queries/CommPkgOperationQuery.cs:55]: Can remove code commented out.

File: src/Equinor.ProCoSys.PcsBus/Queries/CommPkgQueryQuery.cs [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        {whereClause}";
+    }
+
+    private static string CreateWhereClause(long? commPkgId, long? documentId, string? plant)
+    {
+        var whereClause = "";
+        if (commPkgId != null && documentId != null && plant != null)
+        {
+            whereClause = $"where er.projectschema = '{plant}' and er.fromelement_id = {commPkgId} and er.toelement_id = {documentId}";
+        }
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.PcsBus/Queries/CommPkgQueryQuery.cs:39]: Should we use a paramterized approach instead to avoid potential issues with sql injection? Not sure. Checking with @krsteen .

File: src/Equinor.ProCoSys.PcsBus/Queries/SwcrQuery.cs [hunk 19]
```diff
@@ -33,11 +33,11 @@ public static string GetQuery(long? swcrId, string plant = null)
...
-            join project p ON p.project_id = sw.project_id
+            left join project p ON p.project_id = sw.project_id
             left join commpkg c ON c.commpkg_id = sw.commpkg_id
             left join library pri ON pri.library_id = sw.priority_id
             left join library sys On sys.library_id = sw.processsystem_id
-            left join library con ON con.library_id =sw.contract_id
+            left join library con ON con.library_id = sw.contract_id
             left join library cs ON cs.library_id = sw.controlsystem_id
             left join library sup ON sup.library_id = sw.supplier_id
             left join library act On act.library_id = sw.action_id
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.PcsBus/Queries/SwcrQuery.cs:43]: Forgot to refactor this query?

File: src/Equinor.ProCoSys.PcsBus/Queries/WorkOrderMilestoneQuery.cs [hunk 20]
```diff
@@ -8,32 +8,50 @@ public class WorkOrderMilestoneQuery
...
+        //    '"", ""SignedByAzureOid"" : ""' ||  regexp_replace(p.azure_oid, '([""\])', '\\\1') ||
+        //    '"", ""LastUpdated"" : ""' || TO_CHAR(emd.last_updated, 'yyyy-mm-dd hh24:mi:ss') ||        
+        //    '""}}' as message
+        //from elementmilestonedate emd
+        //    join wo on wo.wo_id = emd.element_id
+        //    join project p on p.project_id = wo.project_id
+        //    join library milestone on milestone.library_id = emd.milestone_id         
+        //    left join person p on p.person_id = emd.signedby_id 
+        //{whereClause}";
     }
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.PcsBus/Queries/WorkOrderMilestoneQuery.cs:52]: Code commented out.

File: src/Equinor.ProCoSys.PcsBus/Queries/WorkOrderQuery.cs [hunk 21]
```diff
@@ -5,85 +5,149 @@ public class WorkOrderQuery
...
+        //    left join library hbc on hbc.library_id = w.holdby_id
+        //    left join library dis on dis.library_id = w.discipline_id
+        //    left join library r on r.library_id = w.WORESPONSIBLE_id
+        //    left join library area on area.library_id = w.area_id
+        //    left join library jsc on jsc.library_id = w.jobstatus_id
+        //    left join library tow on tow.library_id = w.typeofwork_id
+        //    left join library osos on osos.library_id = w.onshoreoffshore_id
+        //    left join library woc on woc.library_id = w.wo_id
+        //{whereClause}";
     }
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.PcsBus/Queries/WorkOrderQuery.cs:152]: Code commented out.

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateOnlyHandler.cs (hunk 12 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateOnlyHandler.cs:17]: we're overriding a parse though? if we can't cast to DateTime, we don't want to fail silently.

File: src/Equinor.ProCoSys.BusSender.Core/Models/BusEventMessage.cs (hunk 10 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.BusSender.Core/Models/BusEventMessage.cs:3]: Its to remove the pragma warning: non-nullable property 'Plant' is uninitialized. Consider declaring the property as nullable.

Normally we would initialize using constructor, and remove the init, but Dapper doesn't like constructors so we have to use initializers.

File: src/Equinor.ProCoSys.BusSender.Core/Models/ActionEvent.cs (hunk 9 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.BusSender.Core/Models/ActionEvent.cs:39]: Not sure what you mean, do you think calling the object just Action would be better?

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs (hunk 14 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs:26]: nope, method not even in use, i'll remove

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs (hunk 15 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/DapperRepository.cs:55]: don't understand :D?

File: src/Equinor.ProCoSys.PcsBus/Queries/CommPkgQueryQuery.cs (hunk 18 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.PcsBus/Queries/CommPkgQueryQuery.cs:39]: I've done a check on plant to avoid that, but that may not be the best approach. 

Please advice @krsteen :)

File: src/Equinor.ProCoSys.PcsBus/Queries/SwcrQuery.cs (hunk 19 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.PcsBus/Queries/SwcrQuery.cs:43]: Yes :D

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/TagDetailsRepository.cs [hunk 22]
```diff
@@ -18,7 +18,7 @@ public TagDetailsRepository(BusSenderServiceContext context, ILogger<TagDetailsR
         _logger = logger;
     }
 
-    public async Task<string> GetDetailsStringByTagId(long tagId)
+    public async Task<string?> GetDetailsStringByTagId(long tagId)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.BusSender.Infrastructure/Repositories/TagDetailsRepository.cs:21]: Why return string nullable? 
As far as I can tell all return values from method has a value?

File: src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateOnlyHandler.cs (hunk 12 above)
Comment (jsoi-eq) [reply src/Equinor.ProCoSys.BusSender.Infrastructure/Handlers/DateOnlyHandler.cs:17]: How about a unit test that shows expected behavior when we have a non-castable object value?
That would express that this scenario has been considered and a conscious descision has been made.

File: src/Equinor.ProCoSys.PcsBus/Queries/CommPkgQueryQuery.cs (hunk 18 above)
Comment (krsteen) [reply src/Equinor.ProCoSys.PcsBus/Queries/CommPkgQueryQuery.cs]: I would change it to parameterized anyways just to be sure. It makes the code cleaner as well?

**Review by trondlindanger — APPROVED** [review]
